PlayerMatcher/Matchmaker/Matchmaker.cs

```diff
@@ -44,49 +44,155 @@ public List<User> ConstructMatch(int gameID, int numPlayers)
                                         select players).Count();
 
                 if (totalGamePlayers < numToTake) numToTake = totalGamePlayers;
-
+                                
                 while (playersInMatch.Count < numToTake)
                 {
-                    if (numTaken != 0) numToTake = numToTake - numTaken;
+                    if (numTaken != 0) numForQuery = numToTake - numTaken;
+                    runs++;
 
                     var GetPlayersQuery =
                         from players in db.Users
                         join ranks in db.Ratings on players.User_ID equals ranks.User_ID
                         where ranks.Game_ID == gameID && ranks.User_Rating > minElo && ranks.User_Rating < maxElo
-                        orderby ranks.User_Rating ascending
+                        orderby Guid.NewGuid()
                         select new
                         {
                             Player = players,
                             EloRating = ranks.User_Rating.GetValueOrDefault()
                         };
-
-                    foreach (var matchPlayer in GetPlayersQuery.Take(numToTake))
+                                        
+                    foreach (var matchPlayer in GetPlayersQuery.Take(totalGamePlayers))
                     {
-                        playersInMatch.Add(matchPlayer.Player);
-                        listOfElos.Add(matchPlayer.EloRating);
-                        numTaken++;
+                        if (!playersInMatch.Contains(matchPlayer.Player) && numTaken < numToTake)
+                        {
+                            playersInMatch.Add(matchPlayer.Player);
+                            listOfElos.Add(matchPlayer.EloRating);
+                            numTaken++;
+                        }
+                        if (runs == 1) break;
+                        if (numTaken == numToTake) break;
                     }
 
-                    if (numTaken == 1)
+                    if (runs == 1)
                     {
                         minElo = listOfElos[0] - window;
                         maxElo = listOfElos[0] + window;
                     }
+                    else if (runs % 3 == 0)
+                    {
+                        minElo = minElo - window * 2;
+                        maxElo = maxElo + window * 2;
+                    }
                     else
                     {
                         minElo = minElo - window;
-                        maxElo = maxElo - window;
+                        maxElo = maxElo + window;
                     }
+                    if (minElo < 0) minElo = 0;                    
                 }
 
                 return playersInMatch;
+
+            }
+
+
+        }
+
+        public List<User> ConstructMatch(int gameID, int numPlayers, bool UseBehaviorScore)
```

Comment:
copy pasta code should be consolidated into an if else. also the behavior doesn't change given the boolean value. please address in next PR.